### ISSUE TITLE
Add a `--config-dir` option to the `tedge` command and daemons.

### ISSUE THREAD

**Issue by didier-wenzek** [thin-edge/thin-edge.io#814]
**Motivation**

All the thin-edge commands use implicitly the configuration files stored under `/etc/tedge`.
This is convenient on a daily basis but this also comes with:

* Surprises: different config files are used with `sudo tedge` vs `tedge`.
   * You need `sudo` just to read the system config.
* Unnecessary code complexity: [tedge_config_location.rs](https://github.com/thin-edge/thin-edge.io/blob/main/crates/common/tedge_config/src/tedge_config_location.rs).
* Unnecessary test complexity:
   * Many tests are running as root. 
   * The tests break the configuration of the system.
   * The tests require clean-up functions.

AC:
* Do not change the daily usage of `sudo tedge` and other daemons: `/etc/tedge` is still the implicit thin-edge configuration directory.
* Make `tedge` (without `sudo`) use the same `/etc/tedge` config files, obviously without write permissions.
* Add a `--config-dir` option to the `tedge` command as well as to the mappers and the agent.
* The argument of this `--config-dir` option is the path to a directory with a `tedge.toml` file and the other configuration files expected by thin-edge.
* The default value for this `--config-dir` option is `/etc/tedge`.

Out of Scope:
* Bridge configuration to be ignored for now.

**Comment by matthiasbeyer** [comment]
> The option might be named --config-dir as one expects a directory.

+1 on this, I would expect that a file would be required for `--config`, so `--config-dir` is IMO the right way to go. Also, implicitely expecting other config files next to the one passed to a `--config` parameter is just unexpected "magic" from a users point of view, so another +1 for me on `--config-dir`! :+1: 

> Unnecessary test complexity:
> * Many tests are running as root.
> * The tests break the configuration of the system.
> * The tests require clean-up functions.

Tests should never be required to run as root, nor should they access the filesystem at all, so :+1: on the motivation here! :smiley:

**Comment by didier-wenzek** [comment]
The agreement is to name `--config-dir` this new option.

**Comment by didier-wenzek** [comment]
There are some points to fix along the way.

* [tedge_config_location.rs](https://github.com/thin-edge/thin-edge.io/blob/main/crates/common/tedge_config/src/tedge_config_location.rs) has to be simplified.
* The following paths are hardcoded in misc places. They must be relative to the `--config-dir` option.
   * `/etc/tedge/sm-plugins`
   * `/etc/tedge/device/inventory.json`
   * `/etc/tedge/operations`
   * The tedge `system.toml` and `mosquitto-conf` are already related to the config location, but this has to be double checked.
* Other paths are also hard coded: for logs and locks. Those ones must be located outside the config dir.
   * The simpler here is to do what has been done for the temporary directory with `tedge config set tmp.path`.
* Two configuration files are a bit trickier. The mosquitto bridge configuration and the sm plugins.
   *  Mosquitto needs to be aware of the bridge configurations created by `tedge connect`.
   * The agent must be granted `sudo` access to the plugins.
   * I would prefer to do nothing special for these. I mean let these files in the `--config-dir` directory. A user who changes the `--config-dir` will have to grant the required access to these files.

**Comment by itsyitsy** [comment]
* Other paths are also hard coded: for logs and locks. Those ones must be located outside the config dir.
   * The simpler here is to do what has been done for the temporary directory with `tedge config set tmp.path`.
Will be created in a separate issue lateron.

**Comment by initard** [comment]
> * Other paths are also hard coded: for logs and locks.

Note that after #790 is merged, the log path will need to be changed in the plugin.

**Comment by PradeepKiruvale** [comment]
> The bit that's still not fully clear to me is how the `--config-dir` option used with `tedge connect` will be passed to the mapper/agent processes started by `tedge connect`.
> 
> It is this `tedge connect` execution where I find something missing. When `tedge connect` starts the daemons like `tedge-agent` and `tedge-mapper`, they need to be started with the same `--config-dir` option, that the `tedge connect` command was executed with. But, `tedge connect` doesn't start these daemon processes directly, but uses `systemctl` commands to start these daemons. But, their service definitions wouldn't have the `--config-dir` option in the executable command. So, `tedge connect` "somehow" needs to instruct `systemct` to use the `--config-dir` option that `tedge connect` received to start these daemons. How will this be done? One option is to have `tedge connect` command dynamically update the service definition files and add this `--config-dir` option to the `ExecStart` command value of these services. But, this is not an extensible solution as tedge is service manager agnostic. Are there some other simpler/cleaner alternatives?
> 
> Having said that, if the end goal of this specific story is to expose this `--config-dir` option **ONLY** to **integration tests** so that they can write configs to non-default locations AND **run the agents/mappers directly with this option**, the above mentioned `tedge connect` enhancement is NOT needed and can be addressed separately. It will be needed only when we use `tedge connect` to start these daemons, which currently the **system tests** do or when we expose this option to the end users as well.

That's a good point. As you mentioned it's not an issue for the integration test as the mapper/agent can be invoked with `--config-dir` option. But when we expose this to `tedge connect` then it will be a problem. So, I feel that the `config-dir` can be set using the `tedge config set` and it will be part of the tedge.toml in the `user/home` or `/etc/tedge` depending on how the `tedge config set` is triggered. This will be a clean solution as we do not need to have one more cli option for the mapper or the agent. The logic for parsing the toml file already exists inside the mapper/agent code. So, parsing one more field might not be an issue.

**Comment by albinsuresh** [comment]
@PradeepKiruvale 
> That's a good point. As you mentioned it's not an issue for the integration test as the mapper/agent can be invoked with `--config-dir` option. But when we expose this to `tedge connect` then it will be a problem. So, I feel that the `config-dir` can be set using the `tedge config set` and it will be part of the tedge.toml in the `user/home` or `/etc/tedge` depending on how the `tedge config set` is triggered. This will be a clean solution as we do not need to have one more cli option for the mapper or the agent. The logic for parsing the toml file already exists inside the mapper/agent code. So, parsing one more field might not be an issue.

We can't add this as a new config key in `tedge.toml` because the whole point of this story was to enable the usage of tedge config from somewhere outside the default locations. So, if we just write this new config key into the existing tedge config locations itself, the purpose is lost. So, it has to be persisted somewhere other than the tedge config file itself.

This is why there was another consideration to keep this `--config-dir` value persisted in `system.toml` file so that tedge utilities and the service managers can read this value from that file. But then, we hit the same problem that we're trying to solve with this issue, which is the hardcoded location of `system.toml` in `/etc/tedge which is what we're trying to avoid. So, persisting this value to another file is not a solution as it's more like a chicken and egg problem.

**Comment by didier-wenzek** [comment]
@albinsuresh

For me, the intent is mostly to ease testing. But we cannot enforce that to be use only for testing. And it can make sense to use a different config dir for evaluation.

The issue about `tedge connect` is real. But I would not update the `tedge` command to run the mappers and the agent with specific parameters. `tedge connect` must use the commands provided by the `system.toml` configuration without any changes. Somehow this is the point of this ticket: avoid hardcoded switches in the code.

To fix this issue with `tedge connect`, I would rather fix what is [still hardcoded in the `tedge connect` command: the service names](https://github.com/thin-edge/thin-edge.io/blob/0249b1369e5cfcd53f6e590ab9aa4c23f0278b58/crates/core/tedge/src/system_services/services.rs#L30). One should should use the `system.toml` to tell which services have to be started for `tedge-mapper-c8y` and `tedge-agent`. One can then use specific `tedge-mapper-c8y-test` and `tedge-agent-test` services.

**Comment by initard** [comment]
@didier-wenzek, 

> * Make `tedge` (without `sudo`) use the same `/etc/tedge` config files, obviously without write permissions.

This makes sense for `sudo tedge config` vs `tedge config` but what should the config path be for `sudo tedge connect` vs `tedge connect`
